README.md
# <a name="title"></a> Kitchen::Linode

A Test Kitchen Driver for Linode.

## <a name="requirements"></a> Requirements

Requires [Test Kitchen](http://kitchen-ci.org) and a Linode account.
```
gem install test-kitchen
```

## <a name="installation"></a> Installation and Setup

Install the gem file:
```
gem install kitchen-linode
```
Please read the [Driver usage][driver_usage] page for more details.

## <a name="config"></a> Configuration

For many of these, you can specify an ID number, a full name, or a partial name that will try to match something in the list but may not match exactly what you want.
```
LINODE_API_KEY      Linode API Key environment variable, default: nil
:username           ssh user name, default: "root"
:password           password for user, default: randomly generated hash
:image              Linux distribution, default: "Debian 8"
:data_center        data center, default: "Atlanta"
:flavor             linode type/amount of RAM, default: "Linode 1024"
:payment_terms      if you happen to have legacy, default: 1
:kernel             Linux kernel, default: "Latest 64 bit"
:private_key_path   Location of your private key file, default: "~/.ssh/id_rsa"
:public_key_path    Location of your public key file, default: "~/.ssh/id_rsa.pub"
:ssh_timeout        ssh timeout, default: 600 (seconds)
:sudo               use sudo, default: True
:port               ssh port, default: 22
```

## <a name="usage"></a> Usage

First, set your Linode API key in an environment variable:
```
$ export LINODE_API_KEY='myrandomkey123123213h123bh12'
```
Then, create a .kitchen.yml file:
```
---
driver:
  name: linode

provisioner:
  name: salt_solo
  formula: vim
  state_top:
    base:
      "*":
        - vim

platforms:
  - name: debian_jessie

suites:
  - name: default
```
then you're ready to run `kitchen test` or `kitchen converge`
```
$ kitchen test
```
If you want to create a second yaml config; one for using Vagrant locally but a second to use the Linode driver when run on your CI server, create a config with a name like `.kitchen-ci.yml`:
```
---
driver:
  name: linode

provisioner:
  name: salt_solo
  formula: vim
  state_top:
    base:
      "*":
        - vim

platforms:
  - name: debian_jessie

suites:
  - name: default
```
Then you can run the second config by changing the KITCHEN_YAML environment variable:
```
$ KITCHEN_YAML="./.kitchen-ci.yml" kitchen test
```
If you want to change any of the default settings, you can do so in the 'platforms' area:
```
...<snip>...
platforms:
  - name: debian_jessie
    driver:
      flavor: 2048
      data_center: Dallas
      kernel: 4.0.2-x86_64-linode56
      image: Debian 7
...<snip>...
```

## <a name="development"></a> Development

* Source hosted at [GitHub][repo]
* Report issues/questions/feature requests on [GitHub Issues][issues]

Pull requests are very welcome! Make sure your patches are well tested.
Ideally create a topic branch for every separate change you make. For
example:

1. Fork the repo
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

## <a name="authors"></a> Authors

Created and maintained by [Brett Taylor][author] (<btaylor@linode.com>)

## <a name="license"></a> License

Apache 2.0 (see [LICENSE][license])


[author]:           https://github.com/ssplatt
[issues]:           https://github.com/ssplatt/kitchen-linode/issues
[license]:          https://github.com/ssplatt/kitchen-linode/blob/master/LICENSE
[repo]:             https://github.com/ssplatt/kitchen-linode
[driver_usage]:     http://docs.kitchen-ci.org/drivers/usage


CHANGELOG.md
## 0.9.0

* added some extra info lines

## 0.8.0

* updated some error handling logic

## 0.7.0

* changed default Distribution to 'Debian 8' to match change upstream

## 0.6.0

* added an info line

## 0.5.0

* added loop to retry initial ssh session on Linode

## 0.4.0

* set default ssh timeout to 600 seconds
* added some more information lines

## 0.3.0

* added option for ssh timeout, set default to 60 seconds

## 0.2.0

* Small updates to configuration logic

## 0.1.0 / Unreleased

* Initial release


spec/kitchen/driver/linode_spec.rb
# Encoding: UTF-8

require_relative '../../spec_helper'
require_relative '../../../lib/kitchen/driver/linode'

require 'logger'
require 'stringio'
require 'rspec'
require 'kitchen'
require 'kitchen/driver/linode'
require 'kitchen/provisioner/dummy'
require 'kitchen/transport/dummy'
require 'kitchen/verifier/dummy'
require 'fog'

describe Kitchen::Driver::Linode do
  let(:logged_output) { StringIO.new }
  let(:logger) { Logger.new(logged_output) }
  let(:config) { Hash.new }
  let(:state) { Hash.new }
  let(:rsa) { File.expand_path('~/.ssh/id_rsa') }
  let(:instance_name) { 'the_thing' }
  let(:transport)     { Kitchen::Transport::Dummy.new }
  let(:platform)      { Kitchen::Platform.new(name: 'fake_platform') }
  let(:driver)        { Kitchen::Driver::Openstack.new(config) }

  let(:instance) do
    double(
      name: instance_name,
      transport: transport,
      logger: logger,
      platform: platform,
      to_str: 'instance'
    )
  end

  let(:driver) { described_class.new(config) }

  before(:each) do
    allow_any_instance_of(described_class).to receive(:instance)
      .and_return(instance)
    allow(File).to receive(:exist?).and_call_original
    allow(File).to receive(:exist?).with(rsa).and_return(true)
  end

  describe '#initialize' do
    context 'overridden options' do
      let(:config) do
        {
          image: 139,
          data_center: 10,
          flavor: 2,
          kernel: 215,
          username: 'someuser',
          server_name: 'thisserver',
          private_key_path: '/path/to/id_rsa',
        }
      end

      it 'uses all the overridden options' do
        drv = driver
        config.each do |k, v|
          expect(drv[k]).to eq(v)
        end
      end

      it 'overrides server name prefix with explicit server name, if given' do
        expect(driver[:server_name]).to eq(config[:server_name])
      end
    end
  end

end

lib/kitchen/driver/linode.rb
# -*- encoding: utf-8 -*-
#
# Author:: Brett Taylor (<btaylor@linode.com>)
#
# Copyright (C) 2015, Brett Taylor
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'kitchen'
require 'fog'
require_relative 'linode_version'

module Kitchen

  module Driver
    # Linode driver for Kitchen.
    #
    # @author Brett Taylor <btaylor@linode.com>
    class Linode < Kitchen::Driver::Base
      kitchen_driver_api_version 2
      plugin_version Kitchen::Driver::LINODE_VERSION
      
      default_config :username, 'root'
      default_config :password, nil
      default_config :image, 140
      default_config :data_center, 4
      default_config :flavor, 1
      default_config :payment_terms, 1
      default_config :ssh_key_name, nil
      default_config :kernel, 138
      
      default_config :sudo, true
      default_config :port, 22
      default_config :ssh_timeout, 600
      
      default_config :private_key, nil
      default_config :private_key_path, "~/.ssh/id_rsa"
      default_config :public_key, nil
      default_config :public_key_path, "~/.ssh/id_rsa.pub"
      
      default_config :api_key, ENV['LINODE_API_KEY']
      
      required_config :api_key

      def create(state)
        # create and boot server
        config_server_name
        
        if state[:server_id]
          info "#{config[:server_name]} (#{state[:server_id]}) already exists."
          return
        end
        
        info("Creating Linode.")
        
        server = create_server
        
        # assign the machine id for reference in other commands
        state[:server_id] = server.id
        state[:hostname] = server.public_ip_address
        info("Linode <#{state[:server_id]}> created.")
        info("Waiting for linode to boot...")
        server.wait_for { ready? }
        info("Linode <#{state[:server_id]}, #{state[:hostname]}> ready.")
        setup_ssh(server, state) if bourne_shell?
      rescue Fog::Errors::Error, Excon::Errors::Error => ex
        raise ActionFailed, ex.message
      end

      def destroy(state)
        return if state[:server_id].nil?
        server = compute.servers.get(state[:server_id])

        server.destroy

        info("Linode <#{state[:server_id]}> destroyed.")
        state.delete(:server_id)
        state.delete(:hostname)
      end
      
      private
      
      def compute
        Fog::Compute.new(:provider => 'Linode', :linode_api_key => config[:api_key])
      end
      
      def create_server
        if config[:password].nil?
          config[:password] = [*('a'..'z'),*('A'..'Z'),*('0'..'9')].shuffle[0,20].join
        end
        
        # set datacenter
        if config[:data_center].is_a? Integer
          data_center = compute.data_centers.find { |dc| dc.id == config[:data_center] }
        else
          data_center = compute.data_centers.find { |dc| dc.location == config[:data_center] }
          if data_center.nil?
            data_center = compute.data_centers.find { |dc| dc.location =~ /#{config[:data_center]}/ }
          end
        end
        if data_center.nil?
          fail(UserError, "No match for data_center: #{config[:data_center]}")
        end
        info "Got data center: #{data_center.location}..."
        
        # set flavor
        if config[:flavor].is_a? Integer
          if config[:flavor] < 1024
            flavor = compute.flavors.find { |f| f.id == config[:flavor] }
          else
            flavor = compute.flavors.find { |f| f.ram == config[:flavor] }
          end
        else
          flavor = compute.flavors.find { |f| f.name == config[:flavor] }
          if flavor.nil?
            flavor = compute.flavors.find { |f| f.name =~ /#{config[:flavor]}/ }
          end
        end
        if flavor.nil?
          fail(UserError, "No match for flavor: #{config[:flavor]}")
        end
        info "Got flavor: #{flavor.name}..."
        
        # set image/distribution
        if config[:image].is_a? Integer
          image = compute.images.find { |i| i.id == config[:image] }
        else
          image = compute.images.find { |i| i.name == config[:image] }
          if image.nil?
            image = compute.images.find { |i| i.name =~ /#{config[:image]}/ }
          end
        end
        if image.nil?
          fail(UserError, "No match for image: #{config[:image]}")
        end
        info "Got image: #{image.name}..."
        
        # set kernel
        if config[:kernel].is_a? Integer
          kernel = compute.kernels.find { |k| k.id == config[:kernel] }
        else
          kernel = compute.kernels.find { |k| k.name == config[:kernel] }
          if kernel.nil?
            kernel = compute.kernels.find { |k| k.name =~ /#{config[:kernel]}/ }
          end
        end
        if kernel.nil?
          fail(UserError, "No match for kernel: #{config[:kernel]}")
        end
        info "Got kernel: #{kernel.name}..."
        
        if config[:private_key_path]
          config[:private_key_path] = File.expand_path(config[:private_key_path])
        end
        if config[:public_key_path]
          config[:public_key_path] = File.expand_path(config[:public_key_path])
        end

        # submit new linode request
        compute.servers.create(
          :data_center => data_center,
          :flavor => flavor, 
          :payment_terms => config[:payment_terms], 
          :name => config[:server_name],
          :image => image,
          :kernel => kernel,
          :username => config[:username],
          :password => config[:password]
        )
      end
      
      def setup_ssh(server, state)
        state[:ssh_key] = config[:private_key_path]
        do_ssh_setup(state, config)
      end

      def do_ssh_setup(state, config)
        info "Setting up SSH access for key <#{config[:public_key_path]}>"
        info "Connecting <#{config[:username]}@#{state[:hostname]}>..."
        ssh = Fog::SSH.new(state[:hostname],
                           config[:username],
                           :password => config[:password],
                           :timeout => config[:ssh_timeout])
        pub_key = open(config[:public_key_path]).read
        @max_interval = 60
        @max_retries = 10
        @retries = 0
        begin
          ssh.run([
            %(mkdir .ssh),
            %(echo "#{pub_key}" >> ~/.ssh/authorized_keys),
            %(passwd -l #{config[:username]})
          ])
        rescue
          @retries ||= 0
          if @retries < @max_retries
            info "Retrying connection..."
            sleep [2**(@retries - 1), @max_interval].min
            @retries += 1
            retry
          else
            raise
          end
        end
        info "Done setting up SSH access."
      end
      
      # Set the proper server name in the config
      def config_server_name
        return if config[:server_name]
        config[:server_name] = "kitchen_linode-#{rand.to_s.split('.')[1]}"
      end
    end
  end
end
